# GitHub Repository RAG

### Retrieval-Augmented Generation (RAG) system for querying and interacting with the contents of GitHub repositories.

---

**Author:** David Fernández Martínez  
**Email:** [david.fernxndez.martinez@gmail.com](mailto:david.fernxndez.martinez@gmail.com)  
**LinkedIn:** [linkedin.com/in/david-fernández-martínez](https://linkedin.com/in/david-fernández-martínez)  
**GitHub:** [github.com/davidfernxndez](https://github.com/davidfernxndez)

---

## 10 · Complete RAG Pipeline Evaluation

Evaluating the complete RAG pipeline is essential to determine whether the system can generate accurate, relevant answers grounded in the information retrieved from the source repository. Unlike retrieval-only evaluation, which focuses on the quality of the retrieved documents, end-to-end evaluation considers both the retrieved context and the final answer generated by the LLM.

This notebook evaluates the developed RAG system using **Ragas and an LLM-as-a-Judge approach**. Two complementary metrics are considered **Faithfulness** and  **Answer Relevancy**.

The evaluation uses a dataset of **20 user queries**, together with the corresponding retrieved contexts and LLM-generated answers. The dataset is stored in `evaluation/complete_rag/evaluation_dataset.json`. Its construction process is reproducible and documented in `evaluation/complete_rag/build_evaluation_dataset.ipynb`.

The notebook introduces the evaluation metrics, applies them to the dataset, and analyses the results to assess the strengths and weaknesses of the developed RAG system. It also discusses the **limitations of LLM-as-a-Judge evaluation**, including the influence of the judge model, the limited size of the evaluation dataset, and the fact that automated scores do not guarantee factual correctness.

### Notebook Objectives

* Explain how Ragas can be used to evaluate a complete RAG pipeline using an LLM-as-a-Judge approach.
* Introduce and apply the Faithfulness and Answer Relevancy metrics.
* Analyse the evaluation results obtained using the [`spark-active-learning`](https://github.com/davidfernxndez/spark-active-learning) repository as the reference knowledge source.
* Discuss the main limitations of the evaluation methodology and the conclusions that can reasonably be drawn from the results.

## Imports and Configuration

In [ ]:
import json
import numpy as np
from pathlib import Path
import asyncio
import json
from IPython.display import Markdown, display

from github_rag.evaluator import RAGEvaluator

## 1. Ragas and an LLM-as-a-Judge Approach

**Ragas** is an open-source framework designed to evaluate Retrieval-Augmented Generation (RAG) systems. It provides a set of metrics to assess different aspects of RAG performance, including the quality of retrieved context and the relevance and faithfulness of generated answers. These metrics help identify weaknesses in the pipeline and guide further improvements.

One of the evaluation approaches supported by Ragas is **LLM-as-a-Judge**, in which a Large Language Model (LLM) acts as an automated evaluator. Instead of relying exclusively on manually annotated answers or exact-match comparisons, the judge model analyses the user query, the retrieved context, and the generated answer to assess its quality according to predefined criteria.

## 2. Evaluation Metrics

This evaluation uses two metrics provided by Ragas: **Faithfulness** and **Answer Relevancy**. Both metrics rely on an LLM-as-a-Judge to assess the quality of the generated answer, but they focus on different aspects of the RAG pipeline.

### 2.1 Faithfulness

**Faithfulness** measures whether the claims made in the generated answer are supported by the retrieved context. Its purpose is to identify answers that introduce unsupported information, even when that information sounds plausible.

Ragas evaluates faithfulness by decomposing the generated answer into individual claims and checking whether each claim can be inferred from the retrieved context. The score is calculated as the proportion of claims supported by the context:

$$
\text{Faithfulness} =
\frac{\text{Supported claims}}{\text{Total claims}}
$$

A score closer to **1** indicates that most or all claims are supported by the retrieved context, whereas a score closer to **0** indicates that many claims lack contextual support.

This metric evaluates the consistency between the answer and the retrieved information; it does not establish whether the retrieved context itself is factually correct or whether the answer provides all the information required.

### 2.2 Answer Relevancy

**Answer Relevancy** measures how relevant the generated answer is to the user's question. It focuses on whether the response addresses the information requested, rather than whether its claims are supported by the retrieved context.

Ragas estimates answer relevancy by using an LLM to generate potential questions based on the answer. These generated questions are then compared with the original user query using semantic similarity. The final score is calculated as the average similarity between the original query and the generated questions.

A score closer to **1** indicates that the answer is highly relevant to the user's question, while a lower score suggests that the response may be incomplete, off-topic, or focused on unrelated information.

Unlike Faithfulness, Answer Relevancy does not directly verify the answer against the retrieved context. Therefore, a response can be relevant to the question while still containing unsupported or incorrect claims.

## 3. Evaluation Methodology Limitations

The evaluation results should be interpreted in light of several limitations related to both the **LLM-as-a-Judge methodology** and the design of the evaluation dataset.

### 3.1 Limitations of LLM-as-a-Judge

Although LLM-as-a-Judge provides a practical way to evaluate natural-language responses automatically, it introduces several potential sources of bias and uncertainty:

* **Judge model dependence:** Evaluation scores depend on the capabilities and behaviour of the LLM used as the judge. Different models may assign different scores to the same answer.
* **Evaluation subjectivity:** Some quality criteria require semantic interpretation, which may lead to inconsistent or imperfect judgements.
* **Limited factual verification:** Faithfulness measures whether claims are supported by the retrieved context, not whether that context is factually correct. Similarly, Answer Relevancy measures alignment with the question rather than factual accuracy.

Consequently, these metrics should be considered automated estimates of answer quality rather than definitive measures of correctness. Manual inspection of representative examples remains valuable for validating the results and identifying errors that the metrics may overlook.

### 3.2 Limitations of the Evaluation Dataset

The evaluation dataset contains only **20 user queries** associated with a single GitHub repository. This limited scope is primarily motivated by the API usage constraints associated with both answer generation and LLM-based evaluation.

This introduces several limitations:

* **Limited sample size:** Twenty queries may not adequately represent the full range of possible user questions, query complexities, and retrieval challenges.
* **Repository-specific evaluation:** The results reflect the behaviour of the system when applied to a particular repository and cannot necessarily be generalised to repositories with different structures, programming languages, or documentation styles.
* **Limited statistical reliability:** A small dataset makes the results more sensitive to individual queries, so the observed scores may not accurately represent performance across a broader range of use cases.

### 3.3 Interpretation of the Results

Given these limitations, the results should be interpreted as an **initial assessment of the developed RAG system**, rather than a comprehensive or statistically conclusive evaluation.

The scores provide useful indications of whether the system generates answers that are relevant to user queries and grounded in the retrieved context. They can also help identify potential weaknesses and guide future improvements.

However, the results do not establish that the system will perform equally well across all queries or GitHub repositories. A more robust evaluation would require a larger and more diverse dataset, tests across multiple repositories, and manual validation of a representative subset of the generated answers.

Despite these limitations, this evaluation provides a useful first step towards systematically assessing the complete RAG pipeline and demonstrates how LLM-as-a-Judge can be incorporated into the development process.

## 4. Complete RAG evaluation procedure

In [ ]:
# Load the evaluation dataset
dataset_path = Path(
    "../evaluation/complete_rag/evaluation_dataset.json"
)

with dataset_path.open("r", encoding="utf-8") as file:
    evaluation_dataset = json.load(file)

print(f"Loaded {len(evaluation_dataset)} queries.")

Loaded 20 queries.


In [ ]:
# Load the RAGEvaluator class that implements Ragas framework
evaluator = RAGEvaluator()

The individual evaluation results for each query are presented below, allowing for a manual inspection of the scores and a qualitative assessment of the system's performance.

In [10]:
async def evaluate_with_retry(
    evaluation_function,
    *,
    max_attempts: int = 3,
    initial_wait: int = 20,
    **kwargs,
) -> float:
    """Evaluate a RAG metric with retries for rate-limit errors.

    Args:
        evaluation_function: Asynchronous evaluation function.
        max_attempts: Maximum number of attempts.
        initial_wait: Initial retry delay in seconds.
        **kwargs: Arguments passed to the evaluation function.

    Returns:
        Evaluation score between 0 and 1.

    Raises:
        Exception: If evaluation fails after the allowed attempts
            or a non-rate-limit error occurs.
    """
    for attempt in range(max_attempts):
        try:
            return await evaluation_function(**kwargs)

        except Exception as exc:
            error_message = str(exc).lower()

            is_rate_limit = (
                "429" in error_message
                or "rate_limit_exceeded" in error_message
                or "tokens per minute" in error_message
            )

            if not is_rate_limit or attempt == max_attempts - 1:
                raise

            wait_seconds = initial_wait * (2 ** attempt)

            print(
                f"Rate limit reached. Retrying in {wait_seconds}s "
                f"(attempt {attempt + 2}/{max_attempts})."
            )

            await asyncio.sleep(wait_seconds)

    raise RuntimeError("Evaluation failed unexpectedly.")


# Evaluation loop
results = []
results_path = Path("ragas_evaluation_results.json")

for index, sample in enumerate(evaluation_dataset, start=1):
    user_input = sample["user_input"]
    retrieved_contexts = sample["retrieved_contexts"]
    answer = sample["response"]

    print(f"\n{'=' * 80}")
    print(f"Query {index}/{len(evaluation_dataset)}")
    print(f"{'=' * 80}")
    print(user_input)
    print()

    display(Markdown(answer))

    result = {
        "user_input": user_input,
        "answer": answer,
        "faithfulness": None,
        "answer_relevancy": None,
        "errors": {},
    }

    # Evaluate Faithfulness
    try:
        result["faithfulness"] = await evaluate_with_retry(
            evaluator.evaluate_faithfulness,
            user_input=user_input,
            response=answer,
            retrieved_contexts=retrieved_contexts,
        )
        print(f"Faithfulness: {result['faithfulness']:.4f}")

    except Exception as exc:
        result["errors"]["faithfulness"] = str(exc)
        print(f"Faithfulness failed: {exc}")

    # Evaluate Answer Relevancy
    try:
        result["answer_relevancy"] = await evaluate_with_retry(
            evaluator.evaluate_answer_relevancy,
            user_input=user_input,
            response=answer,
        )
        print(f"Answer relevancy: {result['answer_relevancy']:.4f}")

    except Exception as exc:
        result["errors"]["answer_relevancy"] = str(exc)
        print(f"Answer relevancy failed: {exc}")

    # Store the result, even if one or both metrics failed
    results.append(result)

    # Save progress after every query
    with results_path.open("w", encoding="utf-8") as file:
        json.dump(results, file, ensure_ascii=False, indent=2)

    print(f"Progress saved: {index}/{len(evaluation_dataset)}")

    # Wait before evaluating the next query
    if index < len(evaluation_dataset):
        print("Waiting 60 seconds...")
        await asyncio.sleep(60)

print(f"\nEvaluation finished. Results saved to: {results_path}")


Query 1/20
What clustering method is used for diversity?



The diversity‑based selection is performed with **Bisecting K‑Means** clustering.  
The `diversity_k_means_selection` routine fits a BisectingKMeans model, assigns each candidate to a cluster, and then picks the instance closest to each cluster’s centroid.  

Sources  
- [AL_methods.py](https://github.com/davidfernxndez/spark-active-learning/blob/main/scripts/AL_methods.py)  
- [main.ipynb](https://github.com/davidfernxndez/spark-active-learning/blob/main/main.ipynb)

Faithfulness: 0.7500
Answer relevancy: 0.7849
Progress saved: 1/20
Waiting 60 seconds...

Query 2/20
Why isn't the dataset included in the repository?



The HIGGS dataset is intentionally omitted from the repository because it is very large. The project’s `data_info.md` explicitly states that the HIGGS datasets are not included in the repository due to their size, and users are expected to generate or download them locally using the provided data‑preparation script.  

**Source:** [data_info.md](https://github.com/davidfernxndez/spark-active-learning/blob/main/data/data_info.md)

Faithfulness: 1.0000
Answer relevancy: 0.7529
Progress saved: 2/20
Waiting 60 seconds...

Query 3/20
What is the classifier trained and evaluated in the solution?



The solution does **not hard‑code a particular algorithm** – it is written to work with any Spark MLlib estimator.  
The `train_and_evaluate()` routine trains the chosen estimator on the current labeled set and evaluates it on the test set using a `MulticlassClassificationEvaluator`. Thus, the classifier is left unspecified and can be swapped for any Spark MLlib model (e.g., LogisticRegression, RandomForest, GBT, etc.) when the code is run.  

> Source: [main.ipynb](https://github.com/davidfernxndez/spark-active-learning/blob/main/main.ipynb)

Faithfulness: 1.0000
Answer relevancy: 0.5684
Progress saved: 3/20
Waiting 60 seconds...

Query 4/20
How is uncertainty-based filtering performed?



Uncertainty‑based filtering is carried out in two main steps:

1. **Compute an uncertainty score for every unlabeled instance**  
   * The training DataFrame is first restricted to the unlabeled set \(U_t\).  
   * A model that has been trained only on the labeled instances is used to obtain the class probability \(P(y=1\mid x)\) for each sample in \(U_t\).  
   * For the binary classification task the uncertainty of a sample \(x\) is calculated as  

   \[
   U(x)=1-2\bigl|P(y=1\mid x)-0.5\bigr|
   \]

   which ranges from 0 (certain) to 1 (maximally uncertain).

2. **Select the most uncertain samples**  
   * Let \(B\) be the active‑learning budget (number of samples to label in the current iteration) and \(\alpha\) a multiplier that controls how many candidates are kept.  
   * Compute  

   \[
   p=\min\!\left(1.0,\;\frac{\alpha\,B}{|U_t|}\right)
   \]

   and determine the \((1-p)\)-quantile of the uncertainty scores.  
   * Spark’s `approxQuantile` (Greenwald‑Khanna algorithm) is used to estimate this quantile with a relative error tolerance \(\epsilon\).  
   * All unlabeled instances whose uncertainty is **greater than or equal to** this quantile threshold are retained as candidates for the next diversity‑based selection phase.

The resulting DataFrame contains the columns `id_sample`, `features`, and `uncertainty` for the selected candidates.

> **Sources**  
> [main.ipynb](https://github.com/davidfernxndez/spark-active-learning/blob/main/main.ipynb)  
> [AL_methods.py](https://github.com/davidfernxndez/spark-active-learning/blob/main/scripts/AL_methods.py)

Faithfulness: 0.9333
Answer relevancy: 0.9299
Progress saved: 4/20
Waiting 60 seconds...

Query 5/20
Which are the conclusions regarding the scalability of the solution?



The experiments show that the active‑learning pipeline scales well:

* **Parallelism** – The solution can exploit parallelism effectively up to **4 cores**, achieving a **sub‑linear Size‑up** growth.  
* **Speed‑up behaviour** – When the dataset size increases, the **Spark overhead is amortised** and the algorithm can use more cores more efficiently, leading to better Speed‑up.  
* **Scalability limits** – While the current implementation scales nicely for the tested core counts, larger data volumes may change how many cores can be utilised effectively, as suggested by the Speed‑up experiment.

These conclusions are drawn from the scalability experiments described in the notebook.  

Source: [main.ipynb](https://github.com/davidfernxndez/spark-active-learning/blob/main/main.ipynb)

Faithfulness: 0.8000
Answer relevancy: 0.3711
Progress saved: 5/20
Waiting 60 seconds...

Query 6/20
How is the instance closest to each cluster centroid selected?



The pipeline first clusters the candidate instances with K‑Means and then, for every cluster, it keeps only the instance that is **closest to that cluster’s centroid**.  
The selection is performed in two steps:

1. **Broadcast the centroids** – the centroids of the \(B\) clusters are broadcast to all workers so that each worker can compute distances locally without shuffling the data.

2. **Map each candidate to its cluster and distance** – the `clustered_candidates_df` DataFrame is converted to an RDD where each element is  
   \[
   ( \text{cluster\_id}, (\text{id\_sample}, \text{squared\_distance}) )
   \]
   The distance is computed with `Vectors.squared_distance(row.features, centers_broadcast.value[row.cluster_id])`.  
   (See the `rdd_mapped` definition in *scripts/AL_methods.py*.)

3. **Reduce by cluster to keep the minimum** – the RDD is reduced by key (`cluster_id`) keeping the tuple with the smallest distance. The resulting key‑value pairs give, for each cluster, the sample ID that is nearest to its centroid. These IDs are then used as the selected instances for labeling.

Thus, the instance closest to each cluster centroid is chosen by broadcasting the centroids, mapping each candidate to its cluster and distance, and reducing by cluster to keep the minimum‑distance instance.  

Sources:  
- [main.ipynb](https://github.com/davidfernxndez/spark-active-learning/blob/main/main.ipynb)  
- [scripts/AL_methods.py](https://github.com/davidfernxndez/spark-active-learning/blob/main/scripts/AL_methods.py)

Faithfulness: 0.7000
Answer relevancy: 0.8582
Progress saved: 6/20
Waiting 60 seconds...

Query 7/20
Which MLlib evaluator is used to define the metric in scripts/AL_methods.py?



The metric is defined using **`MulticlassClassificationEvaluator`** from Spark MLlib.  

Source: [scripts/AL_methods.py](https://github.com/davidfernxndez/spark-active-learning/blob/main/scripts/AL_methods.py)

Faithfulness: 1.0000
Answer relevancy: 0.6798
Progress saved: 7/20
Waiting 60 seconds...

Query 8/20
How many Active Learning iterations are executed in the scalability experiment of the script AL_scalability.py?



The scalability experiment in **AL_scalability.py** runs **a single Active Learning iteration**.  
This is explicitly stated in the script’s documentation:

> “The experiment loads and preprocesses the specified dataset, initializes the labeled and unlabeled training sets, and applies **one iteration** of the selected instance sampling strategy.”  

The code itself does not contain a loop over multiple iterations; it starts a timer, trains the model, performs the selection, and then stops the timer once that single iteration is complete.  

**Source:** [scripts/AL_scalability.py](https://github.com/davidfernxndez/spark-active-learning/blob/main/scripts/AL_scalability.py)

Faithfulness: 1.0000
Answer relevancy: 0.9440
Progress saved: 8/20
Waiting 60 seconds...

Query 9/20
How could the performance results of Active Learning be improved?



The repository’s own discussion points out that the modest gains observed in the current experiments are largely a consequence of the **simple experimental setup**: a single‑class, balanced dataset and a basic LogisticRegression model.  To push the performance of the Active Learning (AL) pipeline further, the following directions—explicitly mentioned in the code and documentation—could be pursued:

| Improvement | Why it helps | Where it’s mentioned |
|-------------|--------------|----------------------|
| **Use more expressive MLlib models** (e.g., RandomForest, Gradient‑Boosted Trees, or even deep‑learning models) | Complex models can capture richer patterns, especially when the data are not linearly separable. | *main.ipynb* (future work note) |
| **Test on datasets with higher complexity** (multiclass, highly imbalanced, noisy) | AL strategies often show larger relative benefits on harder problems where labeling is truly expensive. | *main.ipynb* (future work note) |
| **Tune the selection strategy** (e.g., adjust uncertainty thresholds, batch size, or diversity weighting) | Fine‑tuning these hyper‑parameters can reduce redundancy in queried samples and focus learning on the most informative instances. | *main.ipynb* (discussion of moderate gains) |
| **Increase the initial labeled fraction or vary the query batch fraction** | A richer starting point or smaller, more focused batches can accelerate convergence. | *config* (variables referenced in *AL_performance.py*) |
| **Incorporate ensemble or committee‑based uncertainty** | Multiple models can provide a more robust estimate of uncertainty, improving sample selection. | Not explicitly in the repo, but a natural extension of the current uncertainty filtering approach. |
| **Leverage distributed computing more aggressively** (e.g., larger cluster, more cores) | Scaling up can reduce training time per iteration, allowing more iterations or larger batch sizes within the same wall‑clock budget. | *AL_scalability.py* (measures time per iteration) |

In short, the performance of the Active Learning framework can be improved by **(1)** moving beyond LogisticRegression to richer MLlib models, **(2)** evaluating on more challenging datasets, and **(3)** fine‑tuning the selection strategy and experimental parameters. These steps are explicitly suggested in the repository’s own future‑work notes.

Sources:  
[main.ipynb](https://github.com/davidfernxndez/spark-active-learning/blob/main/main.ipynb)  
[README.md](https://github.com/davidfernxndez/spark-active-learning/blob/main/README.md)

Faithfulness: 0.2800
Answer relevancy: 0.7367
Progress saved: 9/20
Waiting 60 seconds...

Query 10/20
How many samples were used for the 100%-sized reference dataset?



The 100 % reference dataset used in the experiments contains **1 000 000 samples**.  
This sample was extracted from the original 11 million‑instance HIGGS dataset to keep the experiments tractable on a local machine.  

Sources  
- [data_info.md](https://github.com/davidfernxndez/spark-active-learning/blob/main/data/data_info.md)  
- [main.ipynb](https://github.com/davidfernxndez/spark-active-learning/blob/main/main.ipynb)

Faithfulness: 1.0000
Answer relevancy: 0.9037
Progress saved: 10/20
Waiting 60 seconds...

Query 11/20
¿What is the dataset used in the experiment



The experiments use the **HIGGS** dataset from the UCI Machine Learning Repository, a binary classification dataset with 11 million instances and 28 numerical features.  

Sources:  
- [main.ipynb](https://github.com/davidfernxndez/spark-active-learning/blob/main/main.ipynb)  
- [data/data_info.md](https://github.com/davidfernxndez/spark-active-learning/blob/main/data/data_info.md)

Faithfulness: 1.0000
Answer relevancy: 0.9502
Progress saved: 11/20
Waiting 60 seconds...

Query 12/20
How is the dataset preprocessed before training the model?



The data that feeds the active‑learning pipeline is prepared in the `load_and_preprocess_data()` function (see **scripts/AL_methods.py**).  
The preprocessing steps are:

1. **Load the raw CSV** – the HIGGS data is read from a file named `higgs‑<percentage_set>.csv` located in the repository’s data directory.  
2. **Cast columns to the required types** – after loading, the columns are cast to the data types expected by the downstream Spark ML pipeline.  
3. **Assemble the feature vector** – all feature columns are combined into a single vector column (`features`) using a `VectorAssembler`.  
4. **Split into training and test sets** – the data is divided into a training set and an independent test set according to the `train_test_split` ratio.  
5. **Standardize the features** – a `StandardScaler` is fitted **only on the training data** and then applied to both the training and test sets. This prevents data leakage.  
6. **Mark initial labels** – within the training set, each instance is randomly assigned a state (`L` for labeled or `U` for unlabeled) based on the `initial_labeled_fraction`. The test set remains untouched by the labeling process.

These steps produce two Spark DataFrames:  
- `train_df` – contains `id_sample`, `label`, standardized `features`, and a `state` column (`L`/`U`).  
- `test_df` – contains `id_sample`, `label`, and standardized `features`.

The resulting DataFrames are then used for the initial model training and subsequent active‑learning iterations.

**Sources**  
- [scripts/AL_methods.py](https://github.com/davidfernxndez/spark-active-learning/blob/main/scripts/AL_methods.py)  
- [README.md](https://github.com/davidfernxndez/spark-active-learning/blob/main/README.md)

Faithfulness: 0.9333
Answer relevancy: 0.6161
Progress saved: 12/20
Waiting 60 seconds...

Query 13/20
How is data leakage prevented during feature standardization?



Data leakage is avoided by **fitting the standard‑scaling model only on the training set** and then re‑using that fitted model to transform both the training and test data.  
In the code:

1. The dataset is split into `train_df` and `test_df` with `df.randomSplit(...)`.  
2. A `StandardScaler` is created and **fitted exclusively on `train_df`** (`scaler.fit(train_df)`), so the mean and standard deviation are computed only from the training data.  
3. The fitted scaler (`scaler_model`) is then applied to both `train_df` and `test_df` (`scaler_model.transform(...)`), ensuring that the test data are scaled using statistics that were not influenced by it.

By never using any information from the test set to compute the scaling parameters, the implementation prevents data leakage during feature standardization.  

Source: [scripts/AL_methods.py](https://github.com/davidfernxndez/spark-active-learning/blob/main/scripts/AL_methods.py)

Faithfulness: 1.0000
Answer relevancy: 0.7840
Progress saved: 13/20
Waiting 60 seconds...

Query 14/20
How is the code of the developed solution structured?



The solution is organized into a small, modular codebase that separates configuration, data handling, and the active‑learning logic:

| Component | Purpose | Key files |
|-----------|---------|-----------|
| **Configuration** | Holds project‑wide constants (paths, dataset URL, random seed, etc.) | `scripts/config.py` |
| **Active‑Learning logic** | Implements the four main stages of the algorithm: data loading/pre‑processing, model training & evaluation, uncertainty filtering, and diversity‑based selection with K‑Means. Each stage is a dedicated function. | `scripts/AL_methods.py` |
| **Orchestration** | Drives the workflow, calls the functions from `AL_methods.py`, and stores results. | `main.ipynb` |

- `scripts/AL_methods.py` contains the following functions:  
  - `load_and_preprocess_data()` – loads the HIGGS dataset and applies the required preprocessing.  
  - `train_and_evaluate()` – trains the model on the labeled set and evaluates it on the test set.  
  - `get_uncertainty_candidates()` – selects instances based on uncertainty.  
  - `diversity_k_means_selection()` – selects diverse instances using K‑Means and simulates expert labeling.  

- `scripts/config.py` defines project paths (`PROJECT_ROOT`, `DATA_DIR`, etc.), dataset parameters (`DATASET_URL`, `TARGET_SAMPLES`, `DATA_PERCENTAGES`), and reproducibility settings (`RANDOM_SEED`).  

- `main.ipynb` imports these modules, runs the pipeline, and records the results, while the README documents the environment and scalability experiments.

**Sources**  
- [main.ipynb](https://github.com/davidfernxndez/spark-active-learning/blob/main/main.ipynb)  
- [scripts/AL_methods.py](https://github.com/davidfernxndez/spark-active-learning/blob/main/scripts/AL_methods.py)  
- [scripts/config.py](https://github.com/davidfernxndez/spark-active-learning/blob/main/scripts/config.py)

Faithfulness: 0.7647
Answer relevancy: 0.6243
Progress saved: 14/20
Waiting 60 seconds...

Query 15/20
Where is broadcast performed in scripts/AL_methods.py?



The broadcast is performed when the cluster centroids are sent to the worker nodes.  
In the `diversity_k_means_selection` routine the centroids computed by K‑Means are broadcast with the Spark context (`sc`) and stored in a variable called `centers_broadcast`.  The workers then access the centroids via `centers_broadcast.value` when mapping the clustered candidates.

> *“sc : pyspark.SparkContext Active Spark context used to broadcast the cluster centroids to the worker nodes.”*  
> *“centers_broadcast.value”*  

Source: [scripts/AL_methods.py](https://github.com/davidfernxndez/spark-active-learning/blob/main/scripts/AL_methods.py)

Faithfulness: 1.0000
Answer relevancy: 0.2881
Progress saved: 15/20
Waiting 60 seconds...

Query 16/20
Which PySpark data structures are used?



The repository uses the following PySpark data structures:

| Data structure | Where it appears | Purpose |
|----------------|------------------|---------|
| **Spark DataFrame** | `train_df`, `test_df` in `AL_methods.py`; `clustered_candidates_df` in the clustering step | Holds tabular data (features, labels, IDs, state) and is the primary API for distributed data manipulation. |
| **RDD** | `clustered_candidates_df.rdd.map(...)` in `AL_methods.py` | Provides low‑level, key‑value pair processing (e.g., mapping each sample to its cluster ID and distance). |
| **Broadcast variable** | `centers_broadcast.value` in `AL_methods.py` | Distributes read‑only data (cluster centroids) efficiently to all worker nodes. |

These structures are explicitly referenced in the README and the `AL_methods.py` script, which describe the distributed active‑learning pipeline.

**Sources**

- [README.md](https://github.com/davidfernxndez/spark-active-learning/blob/main/README.md)  
- [scripts/AL_methods.py](https://github.com/davidfernxndez/spark-active-learning/blob/main/scripts/AL_methods.py)

Faithfulness: 1.0000
Answer relevancy: 0.8649
Progress saved: 16/20
Waiting 60 seconds...

Query 17/20
Explain the config variables defined in scripts/config.py



**Configuration variables in `scripts/config.py`**

| Variable | Purpose | Key details |
|----------|---------|-------------|
| `PROJECT_ROOT` | Base directory of the repository | Determined by resolving the parent of the `scripts` folder (`Path(__file__).resolve().parent.parent`). |
| `DATA_DIR` | Directory where raw data files are stored | `PROJECT_ROOT / "data"`. |
| `COMPRESSED_FILE_PATH` | Path to the compressed HIGGS dataset file | `DATA_DIR / "HIGGS.csv.gz"`. |
| `SCRIPTS_DIR` | Directory containing the project’s scripts | `PROJECT_ROOT / "scripts"`. |
| `RESULTS_DIR` | Directory where experiment results are written | `PROJECT_ROOT / "results"`. |
| `DATASET_URL` | Remote URL to download the HIGGS dataset | `"https://archive.ics.uci.edu/ml/machine-learning-databases/00280/HIGGS.csv.gz"`. |
| `TARGET_SAMPLES` | Desired number of samples to keep from the full dataset | `1_000_000`. |
| `DATA_PERCENTAGES` | List of percentages of the dataset to use in experiments | `[10, 20, 40, 80, 100]`. |
| `RANDOM_SEED` | Seed for all random number generators to ensure reproducibility | `42`. (Defined twice in the file.) |
| `QUERY_BATCH_FRACTION` | Fraction of the training set that will be queried in each active‑learning iteration | `0.01` (i.e., 1 % of the training data). |
| `UNCERTAINTY_CANDIDATE_FACTOR` | Multiplier that determines how many uncertain samples are kept before diversity sampling | `10.0`. It sets the quantile threshold `q = 1 - p` where `p = min(1.0, (UNCERTAINTY_CANDIDATE_FACTOR * B / U_t))`. |
| `UNCERTAINTY_QUANTILE_EPSILON` | Tolerance for approximate quantile estimation used in candidate filtering | `0.001`. |
| `TRAIN_TEST_SPLIT` | Proportion of the data used for training vs. testing | `[0.7, 0.3]` (70 % train, 30 % test). |
| `INITIAL_LABELED_FRACTION` | Fraction of the training data that is initially labeled | `0.05` (5 %). The remaining `95 %` starts as unlabeled. |

These variables configure paths, dataset handling, reproducibility, and the active‑learning workflow (budget, candidate filtering, and initial labeling).  

**Sources**  
[config.py](https://github.com/davidfernxndez/spark-active-learning/blob/main/scripts/config.py)

Faithfulness: 1.0000
Answer relevancy: 0.3494
Progress saved: 17/20
Waiting 60 seconds...

Query 18/20
Is the proposed solution scalable?



Yes.  
The solution is designed to run on Apache Spark and uses only distributed‑friendly operations (e.g., `approxQuantile`, `BisectingKMeans`, and DataFrame filtering).  The authors explicitly report that the algorithm scales well: it exploits parallelism up to four cores, shows sub‑linear size‑up growth, and larger datasets allow Spark overhead to be amortized, enabling more efficient use of additional cores.  Moreover, the scalability experiments script (`AL_scalability.py`) is built to vary the number of cores and dataset size, confirming that the implementation behaves predictably as resources grow.  

**Sources**  
- [main.ipynb](https://github.com/davidfernxndez/spark-active-learning/blob/main/main.ipynb)  
- [scalability_experiments.ipynb](https://github.com/davidfernxndez/spark-active-learning/blob/main/scalability_experiments.ipynb)

Faithfulness: 0.7000
Answer relevancy: 0.5893
Progress saved: 18/20
Waiting 60 seconds...

Query 19/20
Define the scalability measures used.



The scalability experiments in the repository quantify how the solution behaves when the workload or the computing resources change. Three standard metrics are used:

| Metric | What it measures | How it is computed |
|--------|------------------|--------------------|
| **Speed‑up** | The relative reduction in execution time when increasing the number of cores (or workers) for a fixed dataset size. | \( \text{Speed‑up} = \frac{T_{\text{1 core}}}{T_{\text{n cores}}} \) |
| **Size‑up** | The ability of the system to handle larger data volumes while keeping the execution time roughly constant. | \( \text{Size‑up} = \frac{T_{\text{fixed cores, larger data}}}{T_{\text{fixed cores, smaller data}}} \) |
| **Scale‑up** | The combined effect of increasing both data size and computing resources, measuring whether the system can keep the execution time stable as both grow. | \( \text{Scale‑up} = \frac{T_{\text{small data, few cores}}}{T_{\text{large data, many cores}}} \) |

These metrics are calculated from the timing measurements recorded in the scalability experiments and are stored in the CSV files `results/speed_up.csv`, `results/size_up.csv`, and `results/scale_up.csv`.  

Sources:  
- [main.ipynb](https://github.com/davidfernxndez/spark-active-learning/blob/main/main.ipynb)  
- [scripts/AL_scalability.py](https://github.com/davidfernxndez/spark-active-learning/blob/main/scripts/AL_scalability.py)

API call failed on attempt 1: Error code: 429 - {'error': {'message': 'Rate limit reached for model `openai/gpt-oss-120b` in organization `org_01m4fwrb5qe2sa5c9k8m4k2ps2` service tier `on_demand` on tokens per day (TPD): Limit 200000, Used 198274, Requested 3896. Please try again in 15m37.44s. Need more tokens? Upgrade to Dev Tier today at https://console.groq.com/settings/billing', 'type': 'tokens', 'code': 'rate_limit_exceeded'}}
Max retries exceeded. Total attempts: 1, Last error: Error code: 429 - {'error': {'message': 'Rate limit reached for model `openai/gpt-oss-120b` in organization `org_01m4fwrb5qe2sa5c9k8m4k2ps2` service tier `on_demand` on tokens per day (TPD): Limit 200000, Used 198274, Requested 3896. Please try again in 15m37.44s. Need more tokens? Upgrade to Dev Tier today at https://console.groq.com/settings/billing', 'type': 'tokens', 'code': 'rate_limit_exceeded'}}


Faithfulness: 0.4000
Rate limit reached. Retrying in 20s (attempt 2/3).


API call failed on attempt 1: Error code: 429 - {'error': {'message': 'Rate limit reached for model `openai/gpt-oss-120b` in organization `org_01m4fwrb5qe2sa5c9k8m4k2ps2` service tier `on_demand` on tokens per day (TPD): Limit 200000, Used 198228, Requested 3091. Please try again in 9m29.808s. Need more tokens? Upgrade to Dev Tier today at https://console.groq.com/settings/billing', 'type': 'tokens', 'code': 'rate_limit_exceeded'}}
Max retries exceeded. Total attempts: 1, Last error: Error code: 429 - {'error': {'message': 'Rate limit reached for model `openai/gpt-oss-120b` in organization `org_01m4fwrb5qe2sa5c9k8m4k2ps2` service tier `on_demand` on tokens per day (TPD): Limit 200000, Used 198228, Requested 3091. Please try again in 9m29.808s. Need more tokens? Upgrade to Dev Tier today at https://console.groq.com/settings/billing', 'type': 'tokens', 'code': 'rate_limit_exceeded'}}


Rate limit reached. Retrying in 40s (attempt 3/3).


API call failed on attempt 1: Error code: 429 - {'error': {'message': 'Rate limit reached for model `openai/gpt-oss-120b` in organization `org_01m4fwrb5qe2sa5c9k8m4k2ps2` service tier `on_demand` on tokens per day (TPD): Limit 200000, Used 198135, Requested 3896. Please try again in 14m37.392s. Need more tokens? Upgrade to Dev Tier today at https://console.groq.com/settings/billing', 'type': 'tokens', 'code': 'rate_limit_exceeded'}}
Max retries exceeded. Total attempts: 1, Last error: Error code: 429 - {'error': {'message': 'Rate limit reached for model `openai/gpt-oss-120b` in organization `org_01m4fwrb5qe2sa5c9k8m4k2ps2` service tier `on_demand` on tokens per day (TPD): Limit 200000, Used 198135, Requested 3896. Please try again in 14m37.392s. Need more tokens? Upgrade to Dev Tier today at https://console.groq.com/settings/billing', 'type': 'tokens', 'code': 'rate_limit_exceeded'}}


Answer relevancy failed: Error code: 429 - {'error': {'message': 'Rate limit reached for model `openai/gpt-oss-120b` in organization `org_01m4fwrb5qe2sa5c9k8m4k2ps2` service tier `on_demand` on tokens per day (TPD): Limit 200000, Used 198135, Requested 3896. Please try again in 14m37.392s. Need more tokens? Upgrade to Dev Tier today at https://console.groq.com/settings/billing', 'type': 'tokens', 'code': 'rate_limit_exceeded'}}
Progress saved: 19/20
Waiting 60 seconds...

Query 20/20
What is the code used to initialize the labeling state?



The snippets you provided do not contain the exact line of code that creates the initial “labeling state” (i.e., the column that marks each training instance as **L** for labeled or **U** for unlabeled).  
In the repository the labeling state is set up in the data‑loading/pre‑processing step (see the `load_and_preprocess_data()` function in *scripts/AL_methods.py*), but that function’s body is not included in the excerpts you shared.  

From the surrounding context we can infer that the state is initialized by:

1. Adding a column (often called `state` or similar) to the training DataFrame.  
2. Setting that column to **“U”** for all rows.  
3. Marking the initial fraction of instances (defined by `INITIAL_LABELED_FRACTION`) as **“L”**.

However, the exact code that performs this initialization is not present in the provided excerpts. If you need the precise implementation, you’ll have to look at the full `load_and_preprocess_data()` function in *scripts/AL_methods.py*.

**Sources**

- [scripts/AL_performance.py](https://github.com/davidfernxndez/spark-active-learning/blob/main/scripts/AL_performance.py)  
- [main.ipynb](https://github.com/davidfernxndez/spark-active-learning/blob/main/main.ipynb)

API call failed on attempt 1: Error code: 429 - {'error': {'message': 'Rate limit reached for model `openai/gpt-oss-120b` in organization `org_01m4fwrb5qe2sa5c9k8m4k2ps2` service tier `on_demand` on tokens per day (TPD): Limit 200000, Used 197996, Requested 2830. Please try again in 5m56.832s. Need more tokens? Upgrade to Dev Tier today at https://console.groq.com/settings/billing', 'type': 'tokens', 'code': 'rate_limit_exceeded'}}
Max retries exceeded. Total attempts: 1, Last error: Error code: 429 - {'error': {'message': 'Rate limit reached for model `openai/gpt-oss-120b` in organization `org_01m4fwrb5qe2sa5c9k8m4k2ps2` service tier `on_demand` on tokens per day (TPD): Limit 200000, Used 197996, Requested 2830. Please try again in 5m56.832s. Need more tokens? Upgrade to Dev Tier today at https://console.groq.com/settings/billing', 'type': 'tokens', 'code': 'rate_limit_exceeded'}}


Rate limit reached. Retrying in 20s (attempt 2/3).


API call failed on attempt 1: Error code: 429 - {'error': {'message': 'Rate limit reached for model `openai/gpt-oss-120b` in organization `org_01m4fwrb5qe2sa5c9k8m4k2ps2` service tier `on_demand` on tokens per day (TPD): Limit 200000, Used 197949, Requested 2830. Please try again in 5m36.528s. Need more tokens? Upgrade to Dev Tier today at https://console.groq.com/settings/billing', 'type': 'tokens', 'code': 'rate_limit_exceeded'}}
Max retries exceeded. Total attempts: 1, Last error: Error code: 429 - {'error': {'message': 'Rate limit reached for model `openai/gpt-oss-120b` in organization `org_01m4fwrb5qe2sa5c9k8m4k2ps2` service tier `on_demand` on tokens per day (TPD): Limit 200000, Used 197949, Requested 2830. Please try again in 5m36.528s. Need more tokens? Upgrade to Dev Tier today at https://console.groq.com/settings/billing', 'type': 'tokens', 'code': 'rate_limit_exceeded'}}


Rate limit reached. Retrying in 40s (attempt 3/3).


API call failed on attempt 1: Error code: 429 - {'error': {'message': 'Rate limit reached for model `openai/gpt-oss-120b` in organization `org_01m4fwrb5qe2sa5c9k8m4k2ps2` service tier `on_demand` on tokens per day (TPD): Limit 200000, Used 197856, Requested 2830. Please try again in 4m56.352s. Need more tokens? Upgrade to Dev Tier today at https://console.groq.com/settings/billing', 'type': 'tokens', 'code': 'rate_limit_exceeded'}}
Max retries exceeded. Total attempts: 1, Last error: Error code: 429 - {'error': {'message': 'Rate limit reached for model `openai/gpt-oss-120b` in organization `org_01m4fwrb5qe2sa5c9k8m4k2ps2` service tier `on_demand` on tokens per day (TPD): Limit 200000, Used 197856, Requested 2830. Please try again in 4m56.352s. Need more tokens? Upgrade to Dev Tier today at https://console.groq.com/settings/billing', 'type': 'tokens', 'code': 'rate_limit_exceeded'}}
API call failed on attempt 1: Error code: 429 - {'error': {'message': 'Rate limit reached for model `open

Faithfulness failed: Error code: 429 - {'error': {'message': 'Rate limit reached for model `openai/gpt-oss-120b` in organization `org_01m4fwrb5qe2sa5c9k8m4k2ps2` service tier `on_demand` on tokens per day (TPD): Limit 200000, Used 197856, Requested 2830. Please try again in 4m56.352s. Need more tokens? Upgrade to Dev Tier today at https://console.groq.com/settings/billing', 'type': 'tokens', 'code': 'rate_limit_exceeded'}}
Rate limit reached. Retrying in 20s (attempt 2/3).


API call failed on attempt 1: Error code: 429 - {'error': {'message': 'Rate limit reached for model `openai/gpt-oss-120b` in organization `org_01m4fwrb5qe2sa5c9k8m4k2ps2` service tier `on_demand` on tokens per day (TPD): Limit 200000, Used 197810, Requested 3030. Please try again in 6m2.88s. Need more tokens? Upgrade to Dev Tier today at https://console.groq.com/settings/billing', 'type': 'tokens', 'code': 'rate_limit_exceeded'}}
Max retries exceeded. Total attempts: 1, Last error: Error code: 429 - {'error': {'message': 'Rate limit reached for model `openai/gpt-oss-120b` in organization `org_01m4fwrb5qe2sa5c9k8m4k2ps2` service tier `on_demand` on tokens per day (TPD): Limit 200000, Used 197810, Requested 3030. Please try again in 6m2.88s. Need more tokens? Upgrade to Dev Tier today at https://console.groq.com/settings/billing', 'type': 'tokens', 'code': 'rate_limit_exceeded'}}


Rate limit reached. Retrying in 40s (attempt 3/3).


API call failed on attempt 1: Error code: 429 - {'error': {'message': 'Rate limit reached for model `openai/gpt-oss-120b` in organization `org_01m4fwrb5qe2sa5c9k8m4k2ps2` service tier `on_demand` on tokens per day (TPD): Limit 200000, Used 197717, Requested 3030. Please try again in 5m22.704s. Need more tokens? Upgrade to Dev Tier today at https://console.groq.com/settings/billing', 'type': 'tokens', 'code': 'rate_limit_exceeded'}}
Max retries exceeded. Total attempts: 1, Last error: Error code: 429 - {'error': {'message': 'Rate limit reached for model `openai/gpt-oss-120b` in organization `org_01m4fwrb5qe2sa5c9k8m4k2ps2` service tier `on_demand` on tokens per day (TPD): Limit 200000, Used 197717, Requested 3030. Please try again in 5m22.704s. Need more tokens? Upgrade to Dev Tier today at https://console.groq.com/settings/billing', 'type': 'tokens', 'code': 'rate_limit_exceeded'}}


Answer relevancy failed: Error code: 429 - {'error': {'message': 'Rate limit reached for model `openai/gpt-oss-120b` in organization `org_01m4fwrb5qe2sa5c9k8m4k2ps2` service tier `on_demand` on tokens per day (TPD): Limit 200000, Used 197717, Requested 3030. Please try again in 5m22.704s. Need more tokens? Upgrade to Dev Tier today at https://console.groq.com/settings/billing', 'type': 'tokens', 'code': 'rate_limit_exceeded'}}
Progress saved: 20/20

Evaluation finished. Results saved to: ragas_evaluation_results.json


## 5. Results discussion

This section presents the average results obtained, along with their standard deviation, and provides a discussion of the system's performance based on these metrics.

In [12]:
# Load evaluation results
with open("ragas_evaluation_results.json", "r", encoding="utf-8") as file:
    results = json.load(file)

# Extract valid scores, ignoring failed evaluations
faithfulness_scores = [
    r["faithfulness"]
    for r in results
    if r.get("faithfulness") is not None
]

answer_relevancy_scores = [
    r["answer_relevancy"]
    for r in results
    if r.get("answer_relevancy") is not None
]

# Calculate mean and standard deviation
def calculate_statistics(scores):
    n = len(scores)

    if n == 0:
        return 0.0, 0.0

    mean = np.mean(scores)
    std = np.std(scores, ddof=1) if n > 1 else 0.0

    return mean, std


faith_mean, faith_std = calculate_statistics(faithfulness_scores)
rel_mean, rel_std = calculate_statistics(answer_relevancy_scores)

# Display summary
summary_md = f"""
### 📊 LLM-as-a-Judge Evaluation Results (N = {len(results)})

| Metric | Valid scores | Score ($\\mu \\pm \\sigma$) |
| :--- | :---: | :---: |
| **Faithfulness** | {len(faithfulness_scores)}/{len(results)} | `{faith_mean:.3f} ± {faith_std:.3f}` |
| **Answer Relevancy** | {len(answer_relevancy_scores)}/{len(results)} | `{rel_mean:.3f} ± {rel_std:.3f}` |
"""

display(Markdown(summary_md))


### 📊 LLM-as-a-Judge Evaluation Results (N = 20)

| Metric | Valid scores | Score ($\mu \pm \sigma$) |
| :--- | :---: | :---: |
| **Faithfulness** | 19/20 | `0.856 ± 0.215` |
| **Answer Relevancy** | 18/20 | `0.700 ± 0.207` |


The evaluation results show a mean **Faithfulness of 0.856 ± 0.215** and a mean **Answer Relevancy of 0.700 ± 0.207**.

The Faithfulness score suggests that the generated answers are generally well grounded in the retrieved context, although the relatively high standard deviation indicates considerable variability across queries. Some answers may contain unsupported claims, highlighting the importance of examining individual results rather than relying exclusively on the average score.

Answer Relevancy is lower, with a mean of 0.700, suggesting that the generated answers do not always align closely with the user's questions according to the evaluation model. However, this result should be interpreted cautiously. **Manual inspection of the individual query results reveals limitations of the LLM-as-a-Judge approach**, particularly when dealing with more general or open-ended questions. In some cases, the generated answer addresses the question correctly and provides useful information, yet receives a relatively low Answer Relevancy score. This may occur because the judge model does not fully recognise the semantic relationship between a broad question and an answer that provides a more detailed explanation.

Overall, these results indicate that the RAG system generally produces answers grounded in the retrieved context, while answer relevance appears to be more challenging to assess consistently. The discrepancies observed during manual inspection also demonstrate that automated evaluation metrics should not be interpreted in isolation. **Combining quantitative metrics with qualitative inspection provides a more reliable assessment of RAG performance**, especially when the evaluation dataset is small and the judge model may misinterpret some responses.